codec/unknownproto/unknown_fields.go

```diff
@@ -0,0 +1,365 @@
+package unknownproto
+
+import (
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"reflect"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	"google.golang.org/protobuf/encoding/protowire"
+
+	"github.com/cosmos/cosmos-sdk/codec/types"
+)
+
+const bit11NonCritical = 1 << 10
+
+type descriptorIface interface {
+	Descriptor() ([]byte, []int)
+}
+
+type Checker struct {
+	// AllowUnknownNonCriticals when set will skip over non-critical fields that are unknown.
+	AllowUnknownNonCriticals bool
+}
+
+func (ckr *Checker) RejectUnknownFields(b []byte, msg proto.Message) error {
```

Comment:
Why is a pointer receiver needed?

Reply:
That’s an idiomatic rule of thumb for methods on structs that I use — no need for copying within recursive calls.